### DISCUSSION TITLE
How to de-duplicate pymilvus inserts

### DISCUSSION THREAD

**Discussion by zhenzi0322** [milvus-io/milvus#19171] [Q&A and General discussion]
How to de-duplicate pymilvus inserts

**Comment by yhmo** [comment]
Milvus doesn't support de-duplication.
When we say duplication, it has two meanings:
1. duplication of unique ID
User should ensure he provides a unique ID for each entitiy. If he can't, we suggest to set the primary key to be "auto-generated" and let milvus to ensure each ID is unique.
Another workaround: use `query(expr="id in[xxx]")` to check whether this id already exists before you call `insert()`. But we don't recommend this way.

2. duplication of vector
A possible workaround: use the vector to call search() before insert(), if the search() return a top0 that distance=0.0 to this vector, that means the vector already exists. But this workaround performance is poor.

**Reply by zhenzi0322** [reply]
I am currently just assigning different ids to each vector.

**Reply by barzan-hayati** [reply]
> Milvus doesn't support de-duplication.

Does this resolved in the newest version of milvus?

**Reply by yhmo** [reply]
@barzan-hayati 

The upsert() interface can avoid duplicate id: https://milvus.io/docs/insert-update-delete.md#Upsert-entities
upsert() is a heavy operation, intensive upsert requests might be risky.

**Reply by barzan-hayati** [reply]
> The upsert() interface can avoid duplicate id

I don't agree with you. [Does Milvus actually perform deduplication?](https://github.com/milvus-io/milvus/issues/37389)
